### PR TITLE
added unit test for NeedsReanalysisOnOptionChange call

### PR DESCRIPTION
**Pull request by heejaechang**
added unit test for solution crawler that verify it runs correct set of analysis after option changes.

and removed unnecessary abstraction (BaseDiagnosticIncrementalAnalyzer) which is added for v1/v2 engine switch.

fix https://github.com/dotnet/roslyn/issues/13914

### CODE REVIEW

**Comment by heejaechang** [comment]
@srivatsn @mavasani can you take a look?

**Comment by heejaechang** [comment]
ping? this is adding new unit test

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs [hunk 1]
```diff
@@ -265,6 +265,23 @@ public class WorkCoordinatorTests
...
+        {
+            using (var workspace = new WorkCoordinatorWorkspace(SolutionCrawler))
+            {
+                var solutionInfo = GetInitialSolutionInfo(workspace);
+                workspace.OnSolutionAdded(solutionInfo);
+                await WaitWaiterAsync(workspace.ExportProvider);
+
+                var worker = await ExecuteOperation(workspace, w => w.Options = w.Options.WithChangedOption(Analyzer.TestOption, false));
+
+                Assert.Equal(10, worker.SyntaxDocumentIds.Count);
```
Comment (mavasani) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: Seems a bit unclear how the option change is affecting the asserts here. Can you add asserts before ExecuteOperation to make it explicit what changed with that invocation?

Comment (heejaechang) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: basically it is saying it touched all files in the solution. test solution here contains 10 files and 2 projects.

Comment (heejaechang) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: @mavasani I changed name of GetInitialSolutionIfo to make it more explicit what solution it returns

**Comment by heejaechang** [comment]
@srivatsn approval for ask mode?

**Pull request merged**